tari-project/tari
# feat: commitment proofs

Description
---
Adds commitment proofs to the wallet.

Closes #6282.

Motivation and Context
---
It may be useful to assert knowledge of the opening of a commitment, and that it binds to at least a given value. Both can be accomplished through the use of a range proof, since the [Bulletproofs+](https://github.com/tari-project/bulletproofs-plus) range proving API supports the use of minimum value promises.

This PR adds a `CreateCommitmentProof` wallet command that produces such a proof. It accepts as arguments a commitment, message, and optional minimum value; and outputs a hex-encoded proof that binds to these arguments. The proof can be verified using the corresponding `VerifyCommitmentProof` wallet command. Both commands check that the commitment is unspent.

How Has This Been Tested?
---
**Tests are still in progress.**

What process can a PR reviewer use to test or verify this change?
---
Confirm that the proof is constructed and verified correctly using the range proving API.

Manually test:
- that a valid proof verifies against the correct commitment, message, and minimum value
- that a valid proof fails to verify against a different commitment
- that a valid proof fails to verify against a different message
- that a valid proof fails against a different minimum value
- that an invalid proof (for example, a truncated or random hex value) fails to verify
- that a proof cannot be created against a spent commitment
- that a proof fails to verify against a spent commitment
